BLSQ/iaso
# IA-2876 upgrade react router dep

Explain what problem this PR is resolving

Related JIRA tickets : IA-2850

## Self proofreading checklist


## Doc
- None yet

## Changes

- Updated react-router version
- Re-organized `routes.js` and `urls.js`:
    -  urls are now a dict with url (string) and params (array of strings). This is to avoid a circular dependency with the new `useParamsObject` hook
    - routes have a new `routerUrl` param that is mostly used to avoid manipulating `baseUrl` which is used everywhere
    - routes' `component` is now and `element` i.e. a "called" component that takes no props
- Add `useParamsObject` hook:
    - Based on `useParams`
    - gets the `'*'` string params and turns it into an object, like the one previously used by all components
    - To generate the dictionary of params, we use `url.params` from `urls.js`. As a result, it will ignore any additional param not listed. 
- Adapt `useProtectedRoutes` and `ProtectedRoute`:
    - use `navigate` instead of `dispatch` to redirect
- Light refactor of `MenuItemComponent` to avoid app crash
- Add temporary `useRouter` hook, to prevent components with `router` as prop to crash 
- Replace all manual redirects in `TopBar`'s with `useGoBack` (which is still a WIP)
- Add utils `makeRedirectionUrl` to use with `navigate` as replacement of `dispatch(redirect())`
- Add `Redirect` component to handle redirections with params, like `url/param/:params`

## How to test

Login and try to move around the app

Expected to work:
- Home page
- Org unit search
- Org unit details
- Redirections, including legacy urls (eg: `/instances/formId/:formId`) 
- Protected routes, i.e access blocked if no permission

Expected to kinda work:
- back arrow in `TopBar` it should do something but:
    - Due to react-router using relative paths by default, some redirections maybe broken
    - It will always behave as if the page had been reloaded/deep-linked
    - -> This will all be fixed in one go with `useGoBack`. The problem is reliably passing a value through the `location`'s state
- Navigate to org unit child from "Children" table. The user will be redirected but the url will have too many params

Expected not to work:
- Navigate from submission to submission details
- Most redirections from a table icon button


## Notes

- This merges into a dev branch taht will be kept in sync with main, and only merged in main when the whole update is done
- The scope of this PR is the setup of routes, urls and redirections (as in `useRedirections`).
- Since we don't merge in main, don't worry about the test. They'll probably get their own PR
